### PR TITLE
boundary monitor

### LINKED ISSUE #282: Regression rate monitor

### PR DESCRIPTION
**Pull request by mmcgurn**
This PR includes changes to allow for boundary processes to output data to an hdf5 file at the boundary faces and closes  #282.  For testing it also includes a new arbitrary source boundary processes.  This pr cannot be merged until a bug fix in petsc is merged.

The petsc MR is https://gitlab.com/petsc/petsc/-/merge_requests/5575

### CODE REVIEW

File: src/boundarySolver/boundarySolver.cpp [hunk 1]
```diff
@@ -145,6 +146,9 @@ void ablate::boundarySolver::BoundarySolver::Setup() {
                 throw std::runtime_error("Isolated cell " + std::to_string(cell) + " cannot be used in BoundarySolver.");
             }
 
+            // set the faceId and increment
+            geom.faceId = cellFaces[0];
```
Comment (owenmylotte) [inline src/boundarySolver/boundarySolver.cpp:150]: Will this give access to all of the boundary face geometry index wise?

File: src/boundarySolver/boundarySolver.cpp [hunk 2]
```diff
@@ -347,127 +370,109 @@ void ablate::boundarySolver::BoundarySolver::RegisterFunction(ablate::boundarySo
...
             for (const auto& stencilInfo : gradientStencils) {
                 // Get the cell geom
                 const PetscFVCellGeom* cg;
-                DMPlexPointLocalRead(dmCell, stencilInfo.cellId, cellGeomArray, &cg) >> checkError;
+                PetscCall(DMPlexPointLocalRead(dmCell, stencilInfo.cellId, cellGeomArray, &cg));
 
                 // Get pointers to the area of interest
                 const PetscScalar *solPt, *auxPt = nullptr;
-                DMPlexPointLocalRead(dm, stencilInfo.cellId, locXArray, &solPt) >> checkError;
+                PetscCall(DMPlexPointLocalRead(dm, stencilInfo.cellId, locXArray, &solPt));
```
Comment (owenmylotte) [inline src/boundarySolver/boundarySolver.cpp:438]: Are we getting rid of >> checkError; ?

File: src/boundarySolver/physics/arbitrarySource.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#include "arbitrarySource.hpp"
```
Comment (owenmylotte) [inline src/boundarySolver/physics/arbitrarySource.cpp:1]: Does this allow for a heat flux boundary condition?

File: src/boundarySolver/physics/sublimation.cpp [hunk 4]
```diff
@@ -204,6 +202,56 @@ PetscErrorCode ablate::boundarySolver::physics::Sublimation::SublimationFunction
...
+    BoundarySolver::ComputeGradientAlongNormal(dim, fg, auxValues[aOff[TEMPERATURE_LOC]], stencilSize, stencilTemperature.data(), stencilWeights, dTdn);
+
+    // compute the effectiveConductivity
+    PetscReal effectiveConductivity = 0.0;
+    if (sublimation->effectiveConductivity.function) {
+        PetscCall(sublimation->effectiveConductivity.function(boundaryValues, auxValues[aOff[TEMPERATURE_LOC]], &effectiveConductivity, sublimation->effectiveConductivity.context.get()));
+    }
+
+    // compute the heat flux
+    PetscReal heatFluxIntoSolid = -dTdn * effectiveConductivity;
```
Comment (owenmylotte) [inline src/boundarySolver/physics/sublimation.cpp:238]: Will I be able to iterate over the boundaryCell pointer to get face information? Current radiation initialization takes in a solver::Range. I was thinking I'd just put the face indexes into a range unless there is a better way to do it?

**Review by owenmylotte — APPROVED** [review]

**Review by jpdening — APPROVED** [review]

File: src/monitors/boundarySolverMonitor.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    // Create a local vector for just the monitor
+    // create a local vector on the boundary solver
+    Vec globalFaceVec;
+    DMGetGlobalVector(faceDm, &globalFaceVec) >> checkError;
+    PetscObjectSetName((PetscObject)globalFaceVec, GetId().c_str()) >> checkError;
+
+    // map to the local face ids in the faceDm
+    IS subPointIs;
+    DMPlexGetSubpointIS(faceDm, &subPointIs) >> checkError;
```
Comment (jpdening) [inline src/monitors/boundarySolverMonitor.cpp:120]: Is subPointIs ever used?

Comment (mmcgurn) [reply src/monitors/boundarySolverMonitor.cpp:120]: Nice catch!  Thank you!

**Pull request merged**